File: BaseSDK/AArch64/Samples/Base/NYX650/AISingleTriggerResultDelayTest/AISingleTriggerResultDelayTest.cpp
#include <iostream>
#ifdef _WIN32
#include <sys/timeb.h>
#else
#include <time.h>
#endif
#include <thread>
#include <fstream>
#include <string>
#include "Scepter_api.h"
#include "Scepter_Morph_api.h"
#include "get_cpu_vmem.h"
using namespace std;

uint64_t getCurrentTime()
{
#ifdef _WIN32
    timeb timeInfo;
    ftime(&timeInfo);
    return (timeInfo.time * 1000 + timeInfo.millitm);
#else
    timespec timeInfo;
    clock_gettime(CLOCK_REALTIME, &timeInfo);
    return (timeInfo.tv_sec * 1000 + timeInfo.tv_nsec/1000000);
#endif
} 

ScStatus SetParams(ScDeviceHandle deviceHandle)
{
	ScStatus status = SC_OTHERS;

	status = scAIModuleSetWorkMode(deviceHandle, AI_CONTINUOUS_RUN_MODE);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scAIModuleSetWorkMode] success, ScStatus(" << status << "). Set AI_CONTINUOUS_RUN_MODE." << endl;
	}
	else
	{
		cout << "[scAIModuleSetWorkMode] fail, ScStatus(" << status << ")." << endl;
		return status;
	}

	status = scAIModuleSetEnabled(deviceHandle, true);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scAIModuleSetEnabled] success, ScStatus(" << status << "). Enable AI module." << endl;
	}
	else
	{
		cout << "[scAIModuleSetEnabled] fail, ScStatus(" << status << ")." << endl;
		return status;
	}

	return SC_OK;
}

int main()
{
	// Must use GUI tool import specific configuration first
	cout << "---AISingleTriggerResultDelayTest---" << endl;
	uint32_t deviceCount;
	ScDeviceInfo* pDeviceListInfo = NULL;
	ScDeviceHandle deviceHandle = 0;
	ScStatus status = SC_OTHERS;

	//Initialize the ScepterSDK.
	status = scInitialize();
	if (status == ScStatus::SC_OK)
	{
		cout << "[scInitialize] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scInitialize] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}

	//Get the count of devices.
	status = scGetDeviceCount(&deviceCount, 3000);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetDeviceCount] success, ScStatus(" << status << "). The device count is " << deviceCount << endl;
	}
	else
	{
		cout << "[scGetDeviceCount] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}

	if (0 == deviceCount)
	{
		cout << "[scGetDeviceCount] scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples." << endl;
		return 1;
	}

	//Get the infomation of devices.
	pDeviceListInfo = new ScDeviceInfo[deviceCount];
	status = scGetDeviceInfoList(deviceCount, pDeviceListInfo);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetDeviceInfoList] success, ScStatus(" << status << ").";
		if (SC_CONNECTABLE != pDeviceListInfo[0].status)
		{
			cout << " The first device [status]: " << pDeviceListInfo[0].status << " does not support connection." << endl;
			delete[] pDeviceListInfo;
			pDeviceListInfo = NULL;
			return 1;
		}
	}
	else
	{
		cout << "[scGetDeviceInfoList] fail, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
		return 1;
	}

	cout << " The first deviceInfo, <serialNumber>: " << pDeviceListInfo[0].serialNumber
		<< ", <ip>: " << pDeviceListInfo[0].ip << ", <status>: " << pDeviceListInfo[0].status << endl;

	status = scOpenDeviceBySN(pDeviceListInfo[0].serialNumber, &deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scOpenDeviceBySN] success, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
	}
	else
	{
		cout << "[scOpenDeviceBySN] fail, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
		return 1;
	}

	// set soft work model
	status = scSetWorkMode(deviceHandle, SC_SOFTWARE_TRIGGER_MODE);
	if (status != ScStatus::SC_OK)
	{
		cout << "scSetWorkMode failed status:" << status << endl;
		return 1;
	}
	int frameRate = 0;
	status = scGetFrameRate(deviceHandle, &frameRate);
	if (status != ScStatus::SC_OK)
	{
		cout << "scGetFrameRate failed status:" << status << endl;
		return 1;
	}
	void*    pBuffer = 0;
	uint16_t nBufferSize = 0;
	// Get algo version
	uint32_t paramID = 0;
	status = scAIModuleGetParam(deviceHandle, paramID, &pBuffer, &nBufferSize);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scAIModuleGetParam] success, ScStatus(" << status << "). The algo's version is: " << (char*)pBuffer << endl;
	}
	else
	{
		cout << "[scAIModuleGetParam] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}
	// Get algo name
	paramID = 1;
	status = scAIModuleGetParam(deviceHandle, paramID, &pBuffer, &nBufferSize);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scAIModuleGetParam] success, ScStatus(" << status << "). The algo's name is: " << (char*)pBuffer << endl;
	}
	else
	{
		cout << "[scAIModuleGetParam] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}

	//Set bSetParams to false. if the params has been initialized by "Parameter initialization file" in ScepterGUITool or by SDK API in "DeviceImportParamInitFile" sample.
	bool bSetParams = true;
	if (bSetParams)
	{
		status = SetParams(deviceHandle);
		if (status == ScStatus::SC_OK)
		{
			cout << "[SetParams] success, ScStatus(" << status << ")." << endl;
		}
		else
		{
			cout << "[SetParams] fail, ScStatus(" << status << ")." << endl;
			return 1;
		}
	}

	//Start the data stream.
	status = scStartStream(deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scStartStream] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scStartStream] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}

	ofstream csvWriter;
	string fileName = "";
	csvWriter.open(fileName.append("AISingleTriggerResultDelayTest.csv"));
	if (!csvWriter.is_open())
	{
		cout << "csv file open failed" << endl;
		return 1;
	}
	csvWriter << "resultIndex,cpu,mem,TotalDelay,ExcludeDelayofExposure" << endl;

	uint64_t endTimestamp = 0;
	uint64_t startTimestamp = 0;
	uint64_t deviceTimestamp = 0;
	uint64_t frameInterval = 0;
	uint64_t frameIntervalNTP = 0;
	int current_pid = GetCurrentPid();
	float cpu = 0;
	float vmem = 0;
	ScAIResult aiResult = { 0 };
	uint32_t number = 100;
	for (uint32_t i = 0; i < number; i++)
	{
		status = scSoftwareTriggerOnce(deviceHandle);
		startTimestamp = getCurrentTime();
		if (status != ScStatus::SC_OK)
		{
			cout << "[scSoftwareTriggerOnce] fail, ScStatus(" << status << ")." << endl;
			continue;
		} 

		status = scAIModuleGetResult(deviceHandle, 1200, &aiResult);
		endTimestamp = getCurrentTime();
		if (status != SC_OK)
		{
			cout << "[scAIModuleGetResult] fail, ScStatus(" << status << ")." << endl;
			continue;
		}

		if (aiResult.resultIndex % 10 == 0)
		{
			cout << "[scAIModuleGetResult],ScStatus(" << status << "). "
				<< "resultIndex:" << aiResult.resultIndex << "  "
				<< "resultTimestamp:" << aiResult.resultTimestamp << endl;
		}

		deviceTimestamp = aiResult.resultTimestamp;
		frameInterval = endTimestamp - startTimestamp;
		frameIntervalNTP = endTimestamp - deviceTimestamp;
		cpu = GetCpuUsageRatio(current_pid);
		vmem = GetMemoryUsage(current_pid);
		if (frameIntervalNTP > 2000)
		{
			csvWriter << aiResult.resultIndex << "," << cpu << "," << vmem << "," << frameInterval <<  ",N/A" << endl;
		}
		else
		{
			csvWriter << aiResult.resultIndex << "," << cpu << "," << vmem << "," << frameInterval << "," << frameIntervalNTP << endl;
		}
		this_thread::sleep_for(chrono::milliseconds(1000 / frameRate));
	}
	csvWriter.close();
	status = scSetWorkMode(deviceHandle, SC_ACTIVE_MODE);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scSetWorkMode] success, ScStatus(" << status << "). Set SC_ACTIVE_MODE." << endl;
	}
	else
	{
		cout << "[scSetWorkMode] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}
	//Close the opened device.
	status = scCloseDevice(&deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scCloseDevice] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scCloseDevice] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}

	status = scShutdown();
	if (status == ScStatus::SC_OK)
	{
		cout << "[scShutdown] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scShutdown] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}
	cout << "---end---" << endl;

	return 0;
}

File: MultilanguageSDK/Python/Samples/NYX650/AISingleTriggerResultDelayTest/AISingleTriggerResultDelayTest.py
from pickle import FALSE, TRUE
import sys
currentPath =  sys.path[0]
pos = currentPath.find('Samples')
libpath = currentPath[:pos]
print(libpath + "API")
sys.path.append(libpath) #absolutely path

from API.ScepterDS_api import *
import time
import psutil
def get_current_pid():
    # 获取当前进程的PID
    current_pid = psutil.Process().pid
    return current_pid

def get_memory_usage(pid):
    # 获取指定PID的内存占用，单位为MB
    process = psutil.Process(pid)
    memory_info = process.memory_info()
    mem_usage = memory_info.rss / (1024 ** 2)  # rss表示已使用的物理内存
    return mem_usage

def get_cpu_usage(pid):
    # 获取指定PID的CPU占用百分比
    process = psutil.Process(pid)
    cpu_percent = process.cpu_percent(interval=0.5)  # 这里interval参数可选，用于统计间隔时间内的CPU使用率
    return cpu_percent
print("---AISingleTriggerResultDelayTest---")
camera = ScepterTofCam()

def setParameters():
    # Set AI module continuous running.
    ret = camera.scAIModuleSetWorkMode(ScAIModuleMode.AI_SINGLE_RUN_MODE)
    if ret == 0:
        print("[scAIModuleSetWorkMode] success, ScStatus({}). Set AI_SINGLE_RUN_MODE.".format(ret))
    else:
        print("[scAIModuleSetWorkMode] fail, ScStatus({})".format(ret))
        return ret

	# Enable AI module.
    ret = camera.scAIModuleSetEnabled(c_bool(True))
    if ret == 0:
        print("[scAIModuleSetEnabled] success, ScStatus({}). Enable AI module.".format(ret))
    else:
        print("[scAIModuleSetEnabled] fail, ScStatus({})".format(ret))
        return ret
    return 0

camera_count = camera.scGetDeviceCount(3000)
print("[scGetDeviceCount] success, ScStatus(0), The device count is {}".format(camera_count))
if camera_count <= 0:
    print("[scGetDeviceCount] scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples.")
    exit(1)

device_info=ScDeviceInfo()

if camera_count > 0:
    ret,device_infolist=camera.scGetDeviceInfoList(camera_count)
    if ret==0:
        device_info = device_infolist[0]
        print("[scGetDeviceInfoList] success, ScStatus({}). The first deviceInfo, <serialNumber>: {} , <ip>: {} , <status>: {}".format(ret, str(device_info.serialNumber.decode()), str(device_info.ip.decode()), str(device_info.status)))
        if  ScConnectStatus.SC_CONNECTABLE.value != device_info.status:
            print(" The first device [status]: {} does not support connection.".format(str(device_info.status)))
            exit(1)
    else:
        print("[scGetDeviceInfoList] fail, ScStatus({})".format(ret))
        exit(1) 
else: 
    print("there are no camera found")
    exit(1)

ret = camera.scOpenDeviceBySN(device_info.serialNumber)
if  ret == 0:
    print('[scOpenDeviceBySN] success ScStatus({}).'.format(str(ret)))
else:
    print('[scOpenDeviceBySN] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = setParameters()
if  ret == 0:
    print('[setParameters] success ScStatus({}).'.format(str(ret)))
else:
    print('[setParameters] fail ScStatus({}).'.format(str(ret)))
    exit(1)

pbuffer = ctypes.c_void_p()
pbufferSize = ctypes.c_uint16(0)
ret = camera.scAIModuleGetParam(paramID = c_uint32(0), pBuffer = pbuffer, bufferSize = pbufferSize)
if ret == 0:
    str_buffer = ctypes.string_at(pbuffer, pbufferSize.value).decode('ascii')
    print("[scAIModuleGetParam] success ScStatus({}). The algo's version is: {}".format(str(ret), str(str_buffer)))
else:
    print('[scAIModuleGetParam] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = camera.scAIModuleGetParam(paramID = c_uint32(1), pBuffer = pbuffer, bufferSize = pbufferSize)
if ret == 0:
    str_buffer = ctypes.string_at(pbuffer, pbufferSize.value).decode('ascii')
    print("[scAIModuleGetParam] success ScStatus({}). The algo's name is: {}".format(str(ret), str(str_buffer)))
else:
    print('[scAIModuleGetParam] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret, frameRate = camera.scGetFrameRate()
if  ret == 0:
    print('[scGetFrameRate] success ScStatus({}). The device frame rate is {}.'.format(str(ret), frameRate))
else:
    print('[scGetFrameRate] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = camera.scSetWorkMode(ScWorkMode.SC_SOFTWARE_TRIGGER_MODE)
if  ret == 0:
    print('[scSetWorkMode] success ScStatus({}). Set SC_SOFTWARE_TRIGGER_MODE.'.format(str(ret)))
else:
    print('[scSetWorkMode] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = camera.scStartStream()
if  ret == 0:
    print('[scStartStream] success ScStatus({}).'.format(str(ret)))
else:
    print('[scStartStream] fail ScStatus({}).'.format(str(ret)))
    exit(1)
curPath = os.getcwd()
filename = curPath + "/AISingleTriggerResultDelayTest.csv"
try:
    file = open(filename, "w")
except IOError:
    print('csv file open failed')
    exit(1)
file.write("resultIndex,cpu,mem,TotalDelay,ExcludeDelayofExposure\n")
#Get the result form AI module.
ResultCount = 100
for i in range(ResultCount):
    # Call the below api to trigger one result, then the result will be sent.
    # If do not call this function, the result will not be sent and the API of scAIModuleGetResult will return timeout fail.
    ret = camera.scSoftwareTriggerOnce()
    startTime = int(time.time()*1000)
    #print("AIModuleGetResult startTime :", startTime)
    if ret != 0:
        print('[scSoftwareTriggerOnce] fail ScStatus({}).'.format(str(ret)))
        continue
    ret, result = camera.scAIModuleGetResult(waitTime = c_uint32(1200))
    endTime = int(time.time()*1000)
    #print("AIModuleGetResult endTime1 :", endTime)
    if ret != 0:
        print('[scAIModuleGetResult] fail ScStatus({}).'.format(str(ret)))
        time.sleep(0.005)
        continue
    exposureEndTime = result.resultTimestamp
    frameIntervalTime = endTime - startTime
    frameIntervalNtpTime = endTime - exposureEndTime
    if result.resultIndex % 10 == 0:
        print('[scAIModuleGetResult] success ScStatus({}). ScAIResult <resultIndex>: {}'.format(str(ret), str(result.resultIndex)))
    current_pid = get_current_pid()
    mem_usage = get_memory_usage(current_pid)
    cpu_usage = get_cpu_usage(current_pid) # block interface, need to adapt with scAIModuleGetResult timeout, now the algorithm defaults timeout is 1 second
    #print(f"当前进程PID: {current_pid}, 内存占用: {mem_usage:.2f} MB,CPU占用: {cpu_usage}%")
    if frameIntervalNtpTime > 2000:
        file.write("{0},{1},{2},{3},N/A\n".format(result.resultIndex, cpu_usage, mem_usage,frameIntervalTime))
    else:
        file.write("{0},{1},{2},{3},{4}\n".format(result.resultIndex, cpu_usage, mem_usage, frameIntervalTime, frameIntervalNtpTime))
    time.sleep(1.2/frameRate)

ret = camera.scSetWorkMode(ScWorkMode.SC_ACTIVE_MODE)
if  ret == 0:
    print('[scSetWorkMode] success ScStatus({}). Set SC_ACTIVE_MODE'.format(str(ret)))
else:
    print('[scSetWorkMode] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = camera.scStopStream()
if  ret == 0:
    print('[scStopStream] success ScStatus({}).'.format(str(ret)))
else:
    print('[scStopStream] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = camera.scCloseDevice()
if  ret == 0:
    print('[scCloseDevice] success ScStatus({}).'.format(str(ret)))
else:
    print('[scCloseDevice] fail ScStatus({}).'.format(str(ret)))
    exit(1)

exit(0)